controllers/flinkcluster_reconciler.go

```diff
@@ -338,30 +340,74 @@ func (reconciler *ClusterReconciler) reconcileJob() (ctrl.Result, error) {
 	var desiredJob = reconciler.desired.Job
 	var observed = reconciler.observed
 	var observedJob = observed.job
+	var observedPrevJob = observed.previousJob
+	var observedPrevJobStatus = observed.cluster.Status.Components.PreviousJob
+	var runningJobs = []string{}
+
+	if reconciler.observed.flinkJobList != nil {
+		for _, job := range reconciler.observed.flinkJobList.Jobs {
+			if job.Status == "RUNNING" {
+				runningJobs = append(runningJobs, job.ID)
+			}
+		}
+	}
+
+	// Stop and clean up previous job.
+	if observedPrevJob != nil {
+		var prevJobName = observedPrevJob.GetName()
+		log.Info(
+			"Deleting previous job resource", "name", prevJobName)
+		var err = reconciler.deleteJob(observedPrevJob)
+		if err != nil {
+			log.Error(err, "Failed to delete previous job", "name", prevJobName)
+			return ctrl.Result{}, err
+		}
+
+		if observedPrevJobStatus != nil && len(observedPrevJobStatus.ID) > 0 {
+			var prevJobID = observedPrevJobStatus.ID
+			log.Info("Canceling previous job", "jobID", prevJobID)
+			var err = reconciler.cancelJob(
+				prevJobID, true /*takeSavepoint*/, true /*isPrevious*/)
+			if err != nil {
+				log.Error(err, "Failed to cancel previous job", "jobID", prevJobID)
+				return requeueResult, nil
+			}
+		}
+	}
 
 	// Create
 	if desiredJob != nil && observedJob == nil {
 		// If the observed Flink job status list is not nil (e.g., emtpy list), it
 		// means Flink REST API server is up and running. It is the source of
 		// truth of whether we can submit a job.
-		if reconciler.observed.flinkJobList != nil {
+		var jobs = reconciler.observed.flinkJobList
```

Comment:
Probably k8s naming convention, I personally feel super confusing about jobs and status interchanging naming style. Might not be something that needs to be change in this PR, but would be great if we can mitigate this confusion in the future.